docs/proposals/20210203-unmanaged-infrastructure.md

```diff
@@ -0,0 +1,170 @@
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Requirements (Optional)](#requirements-optional)
+      - [Functional Requirements](#functional-requirements)
+        - [FR1](#fr1)
+        - [FR2](#fr2)
+      - [Non-Functional Requirements](#non-functional-requirements)
+        - [NFR1](#nfr1)
+        - [NFR2](#nfr2)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+### Managed cluster infrastructure
+
+Cluster infratructure which lifecycle is managed by a provider infraCluster CR.
+E.g in AWS:
+- Network
+    - VPC
+    - Subnets
+    - Internet gateways
+    - Nat gateways
+    - Route tables
+- Security groups
+- Load balancers
+
+### Unmanaged cluster infrastructure
+Cluster infratructure which lifecycle is not managed by CAPI but rather by an external entity.
+
+### Data plane
+Kubernetes endpoint accesible by a workload cluster admin which nodes capacity is intended exclusively to run user facing workloads
+
+## Summary
+
+This proposal introduces first class support for "unmanaged" infrastructure for CAPI providers to consolidate the boundaries between managed and unmanaged cluster infrastructure.
+
+## Motivation
+
+Currently CAPI providers support an opinionated happy path to create and manage cluster infrastructure lifecycle. The fundamental use case we want to support is bring your own infrastructure. An "unmanaged" CAPI infraCluster won't reconcile or manage the lifecycle of the cluster infrastructure, but CAPI will be able to create compute nodes within it.
+
+This will ease adoption of CAPI in heterogeneous real world environments with restricted privileges and where the provider infrastructure for the cluster needs to be managed out of band.
+
+### Goals
+
+- Introduce support for "unmanaged" infrastructure consistently across CAPI providers.
+- The machine controller must be able to operate and manage machines when the infastructure is "unmanaged".
+
+### Non-Goals/Future Work
+
+- Modify existing managed behaviour.
+
+## Proposal
+
+Providers will add a `ManagementPolicy=Managed|Unmanaged` field to the provider infraCluster spec.
```

Comment:
Does this have to be a field? What about having an annotation exposed in Cluster API that every provider can act upon as a filter? 

Let's take AWS as an example, if a user would want to use something external to manage the infrastructure, we could require the interface is fields within `AWSCluster`. The `AWSMachine` controller, would keep using the existing code to get information from an `AWSCluster`, and doesn't have to know if it's managed internally or externally, it doesn't really matter at that point, the most important thing is that the struct could function as a bridge or interface between the infrastructure resources and the machine controller required data.

Reply:
>Does this have to be a field? What about having an annotation exposed in Cluster API that every provider can act upon as a filter?

A field seems the most reasonable to me as this is a consumer choice, express consumer intent and dictates behaviour, this is not metadata.

>Let's take AWS as an example, if a user would want to use something external to manage the infrastructure, we could require the interface is fields within AWSCluster. The AWSMachine controller, would keep using the existing code to get information from an AWSCluster, and doesn't have to know if it's managed internally or externally, it doesn't really matter at that point, the most important thing is that the struct could function as a bridge or interface between the infrastructure resources and the machine controller required data.

This is no go for scenarios where kubernetes bootstrapping and the cluster infrastructure for the workload are owned by different personas. See "Story 1 - Alternate control plane provisioning with user managed infrastructure".

Also this go be a no go for infra topology not matching managed expectations.

Reply:
> The AWSMachine controller, would keep using the existing code to get information from an AWSCluster, and doesn't have to know if it's managed internally or externally,

I don't think the AWSMachine controller is the intended consumer of this field, I think it's actually the AWSCluster controller.

We had some discussion on monday around the fact that it is actually useful to keep the InfraCluster resource as the source of information for the InfraMachine. If I understood your suggestion correctly (here and in the other thread) the idea is that we put the infracluster information into something other than the infracluster CRD? Thus eliminating the InfraCluster controller from the equation?

@randomvariable pointed out that the AWSCluster controller has some validation code in it that verifies some preconditions for the cloud-controller-manager to work as expected. We would likely still want to keep this validation so that we can ensure the infrastructure provided BYO is still valid. I don't know if other providers have this kind of validation but if they do it's useful to keep the InfraCluster controllers in the loop just for the validation

Reply:
Might be I'm a little bit naive here, but if the cluster infrastructure is externally managed, why the user is required to define an infraCluster. Should not be enough to fill in the ClusterEndpoint, and possibly failure domains?

Also, don't you think that re-using the existing infrastructureCluster objects to map externally managed infrastructure objects could be confusing from a user PoV (no matter of field or annotation)? Could we make it clear introducing new CRDs for externallManaged solutions, if required?

Reply:
> If I understood your suggestion correctly (here and in the other thread) the idea is that we put the infracluster information into something other than the infracluster CRD? Thus eliminating the InfraCluster controller from the equation?

The original thinking I was going through was that if we want to allow for other systems or tools to manage the cloud infrastructure, we should provide a testable and validated "interface" (in this case it'd be the AWSCluster, or InfraCluster if you want to keep things generic). 

With a simple annotation on any InfraCluster, e.g. `cluster.x-k8s.io/managed-by: "<name-of-system>"` we communicate the controllers that they shouldn't perform actions or reconcile these resources. Other controllers wouldn't look at the annotation, but instead only look at the information needed to proceed. For example, a contract requires the FailureDomains to be exposed, another (internal to AWS) may require VPC and security groups, those controllers use the InfraCluster managed by some external entity to reconcile their own resources.

The InfraCluster effectively becomes the tested and validated bridge between different components.

Reply:
> Should not be enough to fill in the ClusterEndpoint, and possibly failure domains?

That tells you nothing about the VPC, subnets or security groups the control plane is in. We need to bare in mind that there are infrastructure specific concerns that are not completely abstractable.